[Refactoring] ConvertConcatenationToInterpolatedString: Support inlineing of other interpolated strings

MaStr11 opened the pull request:
Fixes #48900
Supersedes #49207

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:
```diff
@@ -154,6 +155,37 @@ private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode
                 }
+                else if (syntaxFacts.IsInterpolatedStringExpression(piece) &&
+                    syntaxFacts.IsVerbatimInterpolatedStringExpression(piece) == isVerbatimStringLiteral)
+                {
```
please add a comment explaining the case this is matching, and what you're going to->from.

MaStr11 replied:
Done in 95928ba

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:
```diff
@@ -154,6 +155,37 @@ private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode
+                        }
+                        else
+                        {
+                            Contract.Fail();
```
either `throw ExceptionUtilities.UnexpectedValue()`; or `Debug.Fail`.

MaStr11 replied:
Done in 95928ba

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:
```diff
@@ -2132,5 +2132,19 @@ public bool IsThrowStatement(SyntaxNode node)
+
+        public bool IsVerbatimInterpolatedStringExpression(SyntaxNode node)
+        {
+            var interpolatedStringExpression = (InterpolatedStringExpressionSyntax)node;
+            return interpolatedStringExpression.StringStartToken.IsKind(SyntaxKind.InterpolatedVerbatimStringStartToken);
```
```suggestion
            return node is InterpolatedStringExpressionSyntax interpolatedString && interpolatedString.StringStartToken.IsKind(SyntaxKind.InterpolatedVerbatimStringStartToken);
```

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb:
```diff
@@ -2402,5 +2402,17 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.LanguageServices
+        Function IsVerbatimInterpolatedStringExpression(node As SyntaxNode) As Boolean Implements ISyntaxFacts.IsVerbatimInterpolatedStringExpression
+            Return False
+        End Function
+
```
```suggestion
```

MaStr11 commented:
@CyrusNajmabadi Thanks for the review. 

In my previous PR I added support for the removal of superfluous parentheses in `Interpolation.Expression`.  I used the existing `CanRemoveParentheses(SemanticModel semanticModel)` extension method. So in case there are potentially removable expressions in the new `InterploatedString`, I requested a new sematicModel and removed the parentheses if possible:

https://github.com/MaStr11/roslyn/blob/b7928515a2f32da7b73803d9ffbe7f452a5eec1d/src/Features/CSharp/Portable/CodeRefactorings/ConvertStringConcatToInterpolated/CSharpConvertStringConcatToInterpolatedRefactoringProvider.cs#L104-L144
```CS
        private static async Task<Document> UpdateDocumentAsync(Document document, BinaryExpressionSyntax binaryExpression, ImmutableArray<ExpressionSyntax> concatParts, CancellationToken cancellationToken)
        {
            var interpolatedStringContent = ConvertConcatPartsToInterpolatedStringContent(concatParts);
            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
            var editor = new SyntaxEditor(root, CSharpSyntaxGenerator.Instance);
            var interpolated = InterpolatedStringExpression(
                Token(SyntaxKind.InterpolatedStringStartToken),
                List(interpolatedStringContent));
            // expressions may have superfluous parenthesis after the conversion to an interpolated string:
            // "a" + (1 + 1) -> $"a{(1 + 1)}"
            // if there is any (InterpolationSyntax.Expression is ParenthesizedExpressionSyntax) we need to
            // recompile and check via parenthesized.CanRemoveParentheses(semanticModel) if it can be removed.
            var anyPotentiallyRemovableParenthesis = interpolated.Contents.Any(IsParenthesizedInterpolation);
            var annotation = new SyntaxAnnotation("InterpolatedStringExpressionWithPotentiallyRemovableParanthesisKind");
            if (anyPotentiallyRemovableParenthesis)
            {
                // annotate "interpolated" because we need to find it after the changes are applied
                interpolated = interpolated.WithAdditionalAnnotations(annotation);
            }
            editor.ReplaceNode(binaryExpression, interpolated);

            if (anyPotentiallyRemovableParenthesis)
            {
                // Take the changed document and go through interpolatedNode.Contents to check if any parenthesis can be removed.
                var newDocument = document.WithSyntaxRoot(editor.GetChangedRoot());
                var newRoot = await newDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
                var interpolatedNode = (InterpolatedStringExpressionSyntax)newRoot.GetAnnotatedNodes(annotation).Single();
                var semanticModel = await newDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
                editor = new SyntaxEditor(newRoot, CSharpSyntaxGenerator.Instance);
                foreach (var content in interpolatedNode.Contents.Where(IsParenthesizedInterpolation).Cast<InterpolationSyntax>())
                {
                    var parenthesized = (ParenthesizedExpressionSyntax)content.Expression;
                    if (parenthesized.CanRemoveParentheses(semanticModel))
                    {
                        editor.ReplaceNode(parenthesized, parenthesized.Expression);
                    }
                }
            }

            return document.WithSyntaxRoot(editor.GetChangedRoot());
        }
```

Do we want to reuse that approach here or not?

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:
```diff
@@ -166,29 +165,18 @@ private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode
-                            content.Add(contentPart);
-                            previousContentWasStringLiteralExpression = false;
+                            // if piece starts with a text and the previous part was a string, merge the two parts (see also above)
+                            var newText = ConcatinateTextToTextNode(generator, content.Last(), contentPart.GetFirstToken().Text);
```
1. spelling: "Concatinate"
2. consider giving an example of what this means.

MaStr11 replied:
Done in 6c75e923 and 2772cef7

CyrusNajmabadi commented on src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs:
```diff
@@ -156,6 +156,29 @@ private static Task<Document> UpdateDocumentAsync(Document document, SyntaxNode
+                        else
+                        {
+                            content.Add(contentPart);
+                        }
+                        // Only the first contentPart can be merged, therefore we set previousContentWasStringLiteralExpression to false
```
```suggestion
                        }
                        
                        // Only the first contentPart can be merged, therefore we set previousContentWasStringLiteralExpression to false
```

CyrusNajmabadi reviewed (approved):
Thanks!

MaStr11 commented:
@CyrusNajmabadi Sorry, but I haven't seen your LGTM before I pushed more commits. The commits are only minor clean-ups. I have one more question though: Should I add some logic to remove superfluous parenthesis? There are some approaches to do that. I implemented one with tests as described here: https://github.com/dotnet/roslyn/pull/49229#issuecomment-724012609. I could do this as part of this PR or a new one.

CyrusNajmabadi commented:
>  I have one more question though: Should I add some logic to remove superfluous parenthesis? 

In general, an easy way to do that is to just add the SimplifierAnnotation to the topmost node you're creating.  We should then go through and clean up those parens.  See if that works for you.  Otherwise, i wouldn't bother doing more here (unless it was super easy).

CyrusNajmabadi commented:
> @CyrusNajmabadi Sorry, but I haven't seen your LGTM before I pushed more commits.

that's fine.  i still review the later commits :)

MaStr11 commented:
> SimplifierAnnotation to the topmost node

The Simplifier worked like a charm, but I needed to place it on each `Expression` of a newly generated `InterploationSyntax` (I also tried to place it on the `InterpolatedStringExpression` but that didn't work).

MaStr11 commented:
@CyrusNajmabadi Tests are green now. I added the simplifier annotation to the expressions that get capsulated in an Interpolation, but only if the expression is a ParenthesisExpression.

aolszowka commented:
I am sorry I am new to this process, where does this go? How does this get mainlined?

CyrusNajmabadi commented:
Thanks!
